### PR TITLE
Dynamic Load/Reload Classes

### PR DESCRIPTION
**Pull request by spennihana**
Hello,

This commit allows classes defined at runtime to be loaded/reloaded. For reloading classes, the GOLD array of Icers in TypeMap needs to be mutated (so that Icer's will be regen'd for the same classname). If Class C is being reloaded, its type id (this is the int that the leader node first assigned C) remains the same, but the Icer associated with this id is invalidated, along with any supporting Class definitions from Weaver's ClassPool.

This commit does away with the need for a getter for Weaver's _pool member (as all manipulation can simply be done within the class itself).

There are at least two potential uses of Weaver.loadDynamic, one of them being loop-fused Rapids calls.

---
This change is [Reviewable](https://reviewable.io/reviews/h2oai/h2o-3/179)

### CODE REVIEW

File: h2o-core/src/main/java/water/TypeMap.java [hunk 1]
```diff
@@ -65,11 +67,11 @@
 
   };
   // Class name -> ID mapping
-  static private final NonBlockingHashMap<String, Integer> MAP = new NonBlockingHashMap<>();
+  public static final NonBlockingHashMap<String, Integer> MAP = new NonBlockingHashMap<>();
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/TypeMap.java]: Can we hide inner state via accessors? They can be class reloading specific

Comment (spennihana) [inline h2o-core/src/main/java/water/TypeMap.java]: oops! These can go back to "static private" as before, any way to revert these changes so I don't hurt the history?

**Comment by spennihana** [comment]
Hi, I re-hid the internal GOLD and MAP fields (since they were unintentionally exposed).

**Comment by mmalohlava** [comment]
Can we merge it into master if tests are passing? @abal5 @tomasnykodym
